contrib/babelfishpg_common/src/collation.c

```diff
@@ -1393,4 +1416,72 @@ get_collation_callbacks(void)
 		collation_callbacks_var.has_ilike_node = &has_ilike_node;
 	}
 	return &collation_callbacks_var;
+}
+
+/*
+ * babelfish_define_type_default_collation - would be used to update default collation of Babelfish data types
+ * to correct tsql collation.
+ */
+Oid
+babelfish_define_type_default_collation(Oid typeNamespace)
+{
+	const char *babelfish_dump_restore = GetConfigOption("babelfishpg_tsql.dump_restore", true, false);
+	/* We should only override the default collation for Babelfish data types. */
+	if (strcmp(get_namespace_name(typeNamespace), "sys") != 0)
+		return DEFAULT_COLLATION_OID;
+
+	/* 
+	 * If upgrade is going on then we should use oid corresponding to 
+	 * babelfishpg_tsql.restored_server_collation_name.
+	 */
+	if ((babelfish_dump_restore && 
+		 strncmp(babelfish_dump_restore, "on", 2) == 0) &&
+		 babelfish_restored_server_collation_name)
+		return get_collation_oid_internal(babelfish_restored_server_collation_name);
+
+	get_server_collation_oid_internal(false); /* set and cache server_collation_oid */
+
+	if (OidIsValid(server_collation_oid))
+		return server_collation_oid;
+
+	return DEFAULT_COLLATION_OID;
+}
+
+PG_FUNCTION_INFO_V1(get_babel_server_collation_oid);
+
+Datum
+get_babel_server_collation_oid(PG_FUNCTION_ARGS)
+{
+	PG_RETURN_OID(get_server_collation_oid_internal(false));
+}
+
+PG_FUNCTION_INFO_V1(babelfish_update_server_collation_name);
+
+/*
+ * babelfish_update_server_collation_name - corresponding to sys.babelfish_update_server_collation_name() function
+ * which would be available and strickly be used during 1.x to 2.3 upgrade.
+ */
+Datum
+babelfish_update_server_collation_name(PG_FUNCTION_ARGS)
+{
+	MemoryContext oldContext;
+	/* If babelfish_restored_server_collation_name is set then use it. */
+	if (babelfish_restored_server_collation_name == NULL)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+				 errmsg("Invalid use of function babelfish_update_server_collation_name is detected.")));
+	}
+
+	if (!is_valid_server_collation_name(babelfish_restored_server_collation_name))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+				 errmsg("Invalid value of babelfishpg_tsql.restored_server_collation_name GUC is detected.")));
+	}
+
+	oldContext = MemoryContextSwitchTo(TopMemoryContext);
+	server_collation_name = pstrdup(babelfish_restored_server_collation_name);
```

Comment:
we should free existing memory for server collation name

Reply:
server_collation_name is static variable.

Reply:
Addressed in latest code.